### DISCUSSION TITLE
Tensor of data.y printed to be of [128] size but why when reshaping it warns me RuntimeError: shape '[1, 128]' is invalid for input of size 28?

### DISCUSSION THREAD

**Discussion by Kevin-shihello-world** [pyg-team/pytorch_geometric#6168] [Q&A] [answered]
I was using GFN from<> on PYG. And this model is quite weird. It seems composed of MLP instead of layers of graphconv and passes messages by using global_mean_pool, global_add_pool, and batch normalization. And my experience needs to use its data.y, which prints to be a Tensor with [128] size, the same as its batch size. But it warns me that 127 is too large for its size which is 28. So I tried to figure out what was going on and in the train_eval code I wrote the following code to see if it means in this PyTorch version I can't get items on a one dimension Tensor by using '[]' tag:
```
python
label = data.y#.view(-1)
#label = torch.randint(0,3,(radidx.shape[0],)).cuda()
label__ = torch.reshape(label, [1,128])
print(label[100].item())
```
And it warned me this
```
python
 File "/GFN/train_eval.py", line 241, in train
    label__ = torch.reshape(label, [1,128])
RuntimeError: shape '[1, 128]' is invalid for input of size 28
```
It's so weird, what could be the problem?
I would appreciate it if anyone could find me an answer and I think it would also help others who meet the same problem.
Thank you

**Comment by rusty1s** [comment, marked as answer]
Can you claridy what you mean by GFN? How does that model look like?

**Reply by Kevin-shihello-world** [reply]
Ok, it's one of the ENZYME dataset's top models(the pOk, it's one of the ENZYME dataset's top models(the paper names"Are Powerful Graph Neural Nets Necessary? A Dissection on Graph Classification" in ICLR 2020). I wanted to test my idea on a model that can be trained fast so I could test it and my idea efficiently, this model set its epoch num for 100 so I chose it. And I find its code from a GitHub site found in paperwithcode. The site is :https://github.com/chentingpc/gfn. And I recommend you try it, I found it really can be trained fast, with less than 10 mins(Though I still thought it mysteriously to do so, I mean to use mostly MLP to deal with a GRAPH problem. Whatever, I've checked its train_eval code and couldn't find any strange thing like a cheating code)it can be trained for 70% test accuracy on ENZYME dataset.aper names"Are Powerful Graph Neural Nets Necessary? A Dissection on Graph Classification" in ICLR 2020). I wanted to test my idea on a model that can be trained fast so I could test it and my idea efficiently, this model set its epoch num for 100 so I chose it. And I find its code from a GitHub site found in paperwithcode. The site is :https://github.com/chentingpc/gfn. And I recommend you try it, I found it really can be trained fast, with less than 10 mins(Though I still thought it mysteriously to do so, I mean to use mostly MLP to deal with a GRAPH problem. Whatever, I've checked its train_eval code and couldn't find any strange thing like a cheating code)it can be trained for 70% test accuracy on ENZYME dataset.

**Reply by Kevin-shihello-world** [reply]
By the way, something very interesting happened. After I wrote the last reply I thought about what led to the GFN being trained so fastly except no grapgconv be used, and I thought it could be its bigger batch size and its batch normalization. So I first tried to train a GNN model with a bigger batch size. I chose the HGP-sl which has a lot of properties I like and find it works. A bigger batch size, even without batch normalization, make the HGP-sl gain 12% of test accuracy over the original one with a batch size of 3. And it couldn't be the memory cost that ceases researchers from using a bigger batch size cause the GPU I was using was just a Nividia k80 and a lot of the other studies according to their papers are using GPUs like v100 and so on.

**Reply by Kevin-shihello-world** [reply]
The more interesting thing is I tried and found that batch normalization also helps with HGP-sl's training, and to lessen contingency, I use stored model parameters to initialize the model in the two experiences instead of using rand-seed for once before I found with rand-seed on the same model with the same training set it get very different output for me.

**Reply by rusty1s** [reply]
Interesting findings :) Yes, `batch_size` can actually have huge effects on model performance. Is your initial question resolved? If you have problems running their code, you might have success asking in https://github.com/chentingpc/gfn directly.

**Reply by Kevin-shihello-world** [reply]
The reason for this problem is really more about PYG, as I finally found it (covering my face.jpg). I was thought it more about the tensor arrangement and PYG workflow and my last question seems to haven't been mentioned on their website so sorry I haven't posted an issue about this on their page. 
I found the key to this question is while training, the PYG.dataloader do not make every data contain the same number of graphs, instead if the num of graphs in a training set is divided batch_size not be 0 it will let the last data to contain fewer graphs. I thought in this case it will resent some of the graphs being trained in this epoch, but it turned out that I was wrong. 
So yes, my initial question has been resolved.

**Reply by rusty1s** [reply]
You can also use `drop_last` in the loader to ensure that every batch holds the same number of graphs.

**Reply by Kevin-shihello-world** [reply]
Thanks. And I tried to set test_acc divided (10 + len(train_loader) ) as what shows on the PYGs' examples and HGP-sl, and only to find the test_acc be much less than the HGP-sl and even GIN with code from an example on PYG. I do not know why the GFN still ranks so high on the ENZYME dataset on paperwithcode and gets its paper to ICLR2020. I thought its writers, as from some judged earthy and hard-working universities won't do things like change the results
